firmware/user/modes/mode_flight.c

```diff
@@ -0,0 +1,861 @@
+
+    flight->mode = FLIGHT_MENU;
+
+    flight->menu = initMenu(fl_title, flightMenuCb);
+    addRowToMenu(flight->menu);
+    addItemToRow(flight->menu, fl_flight_exlore);
+    addItemToRow(flight->menu, fl_flight_timed);
+    addRowToMenu(flight->menu);
+    addItemToRow(flight->menu, fl_quit);
+    drawMenu(flight->menu);
+
+    timerDisarm(&(flight->updateTimer));
+    timerSetFn(&(flight->updateTimer), flightUpdate, NULL);
+    timerArm(&(flight->updateTimer), FLIGHT_UPDATE_MS, true);
+    enableDebounce(false);
+}
+
+/**
+ * Called when flight is exited
+ */
+void ICACHE_FLASH_ATTR flightExitMode(void)
+{
+    timerDisarm(&(flight->updateTimer));
+    timerFlush();
+    deinitMenu(flight->menu);
+    os_free(flight);
+}
+
+/**
+ * TODO
+ *
+ * @param menuItem
+ */
+static void ICACHE_FLASH_ATTR flightMenuCb(const char* menuItem)
+{
+    if(fl_flight_exlore == menuItem)
+    {
+        flightStartGame(FL_EXPLORE);
+    }
+    else if (fl_flight_timed == menuItem)
+    {
+        flightStartGame(FL_TIMED);
+    }
+    else if (fl_quit == menuItem)
+    {
+        switchToSwadgeMode(0);
+    }
+}
+
+/**
+ * TODO
+ *
+ * @param type
+ * @param type
+ * @param difficulty
+ */
+static void ICACHE_FLASH_ATTR flightStartGame(flGameType type)
+{
+    flight->mode = FLIGHT_GAME;
+    flight->type = type;
+    flight->frames = 0;
+}
+
+/**
+ * @brief called on a timer, updates the game state
+ *
+ * @param arg
+ */
+static void ICACHE_FLASH_ATTR flightUpdate(void* arg __attribute__((unused)))
+{
+    switch(flight->mode)
+    {
+        default:
+        case FLIGHT_MENU:
+        {
+            drawMenu(flight->menu);
+            break;
+        }
+        case FLIGHT_GAME:
+        {
+            // Increment the frame count
+            flight->frames++;
+
+            flightGameUpdate( flight );
+        }
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+//From https://github.com/cnlohr/channel3/blob/master/user/3d.c
+
+static uint8_t sintable[128] = { 0, 6, 12, 18, 25, 31, 37, 43, 49, 55, 62, 68, 74, 80, 86, 91, 97, 103, 109, 114, 120, 125, 131, 136, 141, 147, 152, 157, 162, 166, 171, 176, 180, 185, 189, 193, 197, 201, 205, 208, 212, 215, 219, 222, 225, 228, 230, 233, 236, 238, 240, 242, 244, 246, 247, 249, 250, 251, 252, 253, 254, 254, 255, 255, 255, 255, 255, 254, 254, 253, 252, 251, 250, 249, 247, 246, 244, 242, 240, 238, 236, 233, 230, 228, 225, 222, 219, 215, 212, 208, 205, 201, 197, 193, 189, 185, 180, 176, 171, 166, 162, 157, 152, 147, 141, 136, 131, 125, 120, 114, 109, 103, 97, 91, 86, 80, 74, 68, 62, 55, 49, 43, 37, 31, 25, 18, 12, 6, };
+
+int16_t ModelviewMatrix[16];
```

Comment:
If possible, put these in `flight_t` so the get alloc'd for this mode, rather than taking up fixed memory all the time.

Reply:
I'll add a const! Is that enough?

Reply:
`const` on `ModelviewMatrix[]` and `ProjectionMatrix[]` don't make too much sense, as those definitely change.

For a little background, last year Bill's maze generator & solver could allocate relatively huge amounts of RAM during execution. The ESP physically had the necessary RAM for the maze, but large chunks of it were statically allocated for global variables used by other modes which were not running at the time, so the allocation would fail. The fix was to have each mode malloc all memory it needs when it starts and free all that memory when it's done. Grouping all variables in a single struct (`flight_t`) keeps that simple.

Ultimately 32 bytes aren't going to break the bank, but I'd like to keep it all together for cleanliness.

Reply:
`sintable[]` is fine as-is. If we reeeeealy wanted the space, we could put it in ROM with `RODATA_ATTR`, but the accesses would have to be 32-bit aligned, and would likely be slower, so meh.